torch/csrc/jit/frontend/ir_emitter.cpp

```diff
@@ -1320,13 +1320,31 @@ struct to_ir {
     pushFrame(comprehension_block);
     WithInsertPoint guard(comprehension_block);
     auto emit_body = [&]() {
-      auto comprehension_out = emitExpr(lc.elt());
+      Value* out = emitExpr(lc.elt());
+
+      // Make sure that any element types are subtypes of the annotated
+      // type
+      std::stringstream ss;
+      if (type_hint && !out->type()->isSubtypeOfExt(type_hint->expect<ListType>()->getElementType(), &ss)) {
+        throw ErrorReport(loc) << ss.str() << "List type annotation did"
+                               << " not match the types of the actual "
+                               << "list items";
+      }
+
+      // If we didn't have a type annotation, let the type begin as the
+      // first element in the list
       if (!type_set) {
```

Comment:
Why is this conditional type inference needed when the next step is to infer unified type unconditionally?

Reply:
If we didn't have a type annotation, the type of the list would be set to `Tensor`. We don't want to unify this default type with the actual elements in the list. I've updated the comment to explain this edge case